### PR TITLE
Issue #825: inline expiry

### PR DESCRIPTION
**Pull request by ljacomet**
* implement inline expiry
* proper return contracts for different methods in the face of inline expiry
* Added SPITests where relevant

### CODE REVIEW

File: impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java [hunk 1]
```diff
@@ -340,7 +327,7 @@ private PutStatus putReturnHolder(final K key, final V value) throws CacheAccess
       if (entryActuallyAdded.get()) {
         enforceCapacity(valuePut.size(), eventSink);
       } else {
-        long replacedDelta = valuePut == null ? 0 : valuePut.size() - (replacedValue.get() == null ? 0 : replacedValue.get().size());
+        long replacedDelta = (valuePut == null ? 0 : valuePut.size()) - (replacedValue.get() == null ? 0 : replacedValue.get().size());
```
Comment (lorban) [inline impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java]: Two ternary operators on the same line is a bit much for my taste. I'd prefer to see this split onto two lines.

File: impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java [hunk 2]
```diff
@@ -1233,30 +1234,35 @@ public V apply(final K k, final V oldValue) {
...
+    if (Duration.ZERO.equals(duration)) {
       expireMapping(key, valueHolder);
       return null;
     }
-    valueHolder.accessed(now, duration);
     return valueHolder;
   }
 
-  private OnHeapValueHolder<V> setAccessTimeAndExpiryThenReturnMapping(K key, OnHeapValueHolder<V> valueHolder, long now,
+  private OnHeapValueHolder<V> setAccessTimeAndExpiryThenReturnMappingUnderLock(K key, OnHeapValueHolder<V> valueHolder, long now,
```
Comment (lorban) [inline impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java]: The UnderLock and OutsideLock variants of this method are identical (practically so). Why having two variants then?

File: impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java (hunk 1 above)
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java]: Agreed, but would like to leave it like that and address it in #851

File: impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java (hunk 2 above)
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java]: While the difference is small, there is a huge impact. Remember that events cannot be fired outside of lock scope. So one variant will just fire an expiry event and return null while the other one will perform a full expiration (`computeIfPresent` with equality check that fires event + returns null).

Comment (lorban) [inline impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java]: Okay, then I would suggest to add a comment around line `expireMapping(key, valueHolder);` / `onExpiration(key, valueHolder, eventSink);` as it is very easy to miss the subtle difference.

Comment (ljacomet) [inline impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java]: Done

**Pull request merged**